clog/logger.go

```diff
@@ -0,0 +1,238 @@
+		zsl: zsl,
+		set: set,
+	}
+
+	return cloggerton
+}
+
+// ------------------------------------------------------------------------------------------------
+// context management
+// ------------------------------------------------------------------------------------------------
+
+type loggingKey string
+
+const ctxKey loggingKey = "clog_logger"
+
+// Init embeds a logger within the context for later retrieval.
+// It is a preferred, but not necessary, initialization step.
+// If you don't call this and you start logging, or you call
+// Singleton(), then the package will initialize a logger instance
+// with the default values.  If you need to configure your logs,
+// make sure to embed this first.
+func Init(ctx context.Context, set Settings) context.Context {
+	clogged := singleton(set)
+	clogged.zsl.Debugw("seeding logger", "logger_settings", set)
+
+	return plantLoggerInCtx(ctx, clogged)
+}
+
+// PlantLogger allows users to embed their own zap.SugaredLogger within the context.
+// It's good for inheriting a logger instance that was generated elsewhere, in case
+// you have a downstream package that wants to clog the code with a different zsl.
+func PlantLogger(ctx context.Context, seed *zap.SugaredLogger) context.Context {
+	return plantLoggerInCtx(ctx, &clogger{zsl: seed})
+}
+
+// plantLoggerInCtx allows users to embed their own zap.SugaredLogger within the
+// context and with the given logger settings.
+func plantLoggerInCtx(
+	ctx context.Context,
+	clogger *clogger,
+) context.Context {
+	if clogger == nil {
+		return ctx
+	}
+
+	return context.WithValue(ctx, ctxKey, clogger)
+}
+
+// fromCtx pulls the clogger out of the context.  If no logger exists in the
+// ctx, it returns the global singleton.
+func fromCtx(ctx context.Context) *clogger {
+	l := ctx.Value(ctxKey)
+	// if l is still nil, we need to grab the global singleton or construct a singleton.
+	if l == nil {
+		l = singleton(Settings{}.EnsureDefaults())
+	}
+
+	return l.(*clogger)
+}
+
+// Ctx retrieves the logger embedded in the context.
+// It also extracts any clues from the ctx and adds all k:v pairs to that log instance.
+func Ctx(ctx context.Context) *builder {
+	return newBuilder(ctx)
+}
+
+// CtxErr is a shorthand for clog.Ctx(ctx).Err(err)
+func CtxErr(ctx context.Context, err error) *builder {
+	nb := newBuilder(ctx)
+	nb.err = err
+
+	return nb
+}
+
+// Singleton is a shorthand for .Ctx(context.Background()).  IE: it'll use the singleton
+// logger directly; building one if necessary.  You should avoid this and use .Ctx or
+// .CtxErr if possible.  Likelihood is that you're somewhere deep in a func chain that
+// doesn't accept a ctx, and you still want to add a quick log; maybe for debugging purposes.
+//
+// That's fine!  Everything should work great.
+//
+// Unless you call this before initialization.  Then it'll panic.  We do want you to init
+// the logger first, else you'll potentially lose these logs due different buffers.
+func Singleton() *builder {
```

Comment:
I'm not really a fan of having something like this. If we run into the situation described, it's probably a design issue that should be addressed. I'm concerned something like this will allow more design issues to slip through because they allow kind of papering over the root problem

Reply:
I'm back and forth on it.  I do agree that not providing this method makes a good forcing function to resolve cases that would sloppily use it.  I'm also aware that every now and then devs will hit a corner case where this sort of accessibility makes life much less painful or arduous.  I'm not in a rush to do so, but we can take it out if there are strong feelings against the design.  That's a two-way-door.

Reply:
Temporal already has logging hooks built in, so there's no concerns with that.  But, sure, I can remove it.

Reply:
see: https://github.com/alcionai/clues/pull/67